A---
title: "_cSEAdb_ User's Guide"
author: |
  | Xiang Liu <xiang.liu@moffitt.org>
  | Mingxiang Teng <mingxiang.teng@moffitt.org>
  | Department of Biostatistics and Bioinformatics
  | Moffit Cancer Center, Tampa, FL, USA
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
  pdf_document:
    toc: yes
graphics: yes
always_allow_html: yes
vignette: |
  %\VignetteIndexEntry{cSEAdb user's guide} %\VignettePackage{cSEAdb} %\VignetteEncoding{UTF-8} %\VignetteEngine{knitr::rmarkdown}
---

```{r, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  error = FALSE,
  warning = FALSE, 
  message = FALSE,
  cache = FALSE,
  comment = "#>",
  dev="png"
)
```

# Introduction
Super enhancers (SE) have been found with cancer-specific gene regulation in human cancers.
Recent studies reported the importance of understanding the cooperation and function of
their internal components, i.e., the constituent enhancers (CE).  Here, we report fingerprint
SE signatures at the constituent enhancer level for 28 cancer types in the NCI-60 cell panel. 

# Install package

Install _cSEAdb_ package via _devtools_.
```{r install}
library(devtools)
devtools::install_github("https://github.com/tenglab/cSEAdb.git")
```

# Getting Started

Load the package in R.

```{r library}
library(cSEAdb)
```

# Read cSEAdb RDS object

We document SE signatures in a RDS obeject for query, exploration and visualization. These signatures
were pre-calcuated based on our computational algorithms detailed in the manuscript.

```{r readRDS}
cSEAdb <- readRDS(system.file("data","cSEAdb.rds",package="cSEAdb"))
names(cSEAdb)
```

# Tables of cSEAdb object

cSEAdb contains the following files:

  1. se_specificity: main fingerprint file including specific SEs and CEs across cancers and cancer cell lines.
  The file includes the following columns:
  
    + se_name: SE names (chr_start_end).
    + number_spec_ce: number of specific CEs for each SE.
    + number_spec_object: number of cells (cancers) of which the CE is specific.
    + spec_ce: CEs that show cell or cancer specificity.
    + spec_object: name of cell line or cancer type associated to the specific CE.
    + specificity: types of cell(cancer) specificity: active, inactive or non-specific.
    + object_type: level of specificity: cell-specific or cancer-specific.
    
  1. se_bed: SE BED file where SEs were merged across NCI-60 cell lines.
  1. ce_bed: CE BED file where CEs were merged across NCI-60 cell lines.
  1. cell_cancer_meta: metadata for cell lines and cancers of NCI-60 cell lines.
  1. ce_signal: normalized H3K27Ac ChIP-seq coverage for all CEs.
  1. ce_mixture_model_cell: mixture group code of all CEs (low signal group: 0; high signal group: 1).
  1. gene_promotor: gene promotor regions based on hg38 to facilitate gene query.
  

# Query with _cSEAdb_

In this section, we demo different types of queries to zoom into specific SEs.

### *Query _cSEAdb_ with SE regions*

Query through a specified genomic coordinate or a bed-file-specified region to find SEs.

```{r Query SE regions}
hits_1 <- search_db(c("chr1_100584859_100673000","chr1_112388326_112392055"),
                     query_type="se_region",se_db=cSEAdb)
head(hits_1)
bed_file <- system.file("extdata","se_query_example.bed",package="cSEAdb")
hits_2 <- search_db(bed_file,query_type="se_bed",se_db=cSEAdb)

head(hits_2)
```

### *Query _cSEAdb_ with gene names*

Query through gene names to find the neighbored SE regions .

```{r Query with gene names}
hits_gene <- search_db(query=c("MYC","FOXA2"),
                        query_type="gene",se_db=cSEAdb)
head(hits_gene)
```

### *Query _cSEAdb_ with cell lines (limited to the 60 cell lines in this study)*

Query through cell line names to find the cell-sepcific SEs in the queried cell lines.

```{r Query with cell lines}
# cell line names included in this study
cell_name <- unique(cSEAdb$cell_cancer_meta$Cell_line)
head(cell_name)
# search cSEAdb with cell names
hits_cell <- search_db(cell_name[c(1,2)],
                       query_type="cell",se_db=cSEAdb)
head(hits_cell)
```

### *Query _cSEAdb_ with cancer types (limited to the 28 cancer types in this study)*

Query through cancer type names to find the cancer-sepcific SEs in the queried cancers.

```{r Query with cancer types}
# cancer types included in this study
cancer_name <- unique(cSEAdb$cell_cancer_meta$Cell_line)
head(cancer_name)
# search cSEAdb with cancer types
hits_cancer <- search_db(query=cancer_name[c(1,2)],
                       query_type="cancer",cSEAdb)
head(hits_cancer)
```

# Generate signal tracks for the queried SE regions

In this section, we demo how to create ChIP-seq H3K27Ac data object in order to visualze SE signals.

First, the normalized bigwig files shoule be downloaded to local computer from our pre-documented repository.

```{r download bigwig files of 60 cell lines}
# Download the normalized bigwig files of NCI-60 cell lines
# Download links are located in data/bw_all_plinks.txt file
bw_links <- system.file("extdata","bw_all_plinks.txt",package="cSEAdb")
bw_download_link <- read.table(bw_links,sep="\t",header=F)
head(bw_download_link)
# Create a bigwig list using the path of bigwig files
# Here is an example of path list for downloaded bigwig files
bw_dir <- system.file("extdata","example_bw",package="cSEAdb")
bw_path  <- list.files(bw_dir,full.names = T)
head(bw_path)
```

Then, a data object (GRanges) can be generated to store the H3K27Ac signal tracks for the queried SE regions.
Although, only cancers/cells in the NCI-60 are allowed to be queried above, visulization of signals of the queried SEs in
other cell lines is allowed here. See below on how to add other bigwig files duing the data object preparation.

```{r create bw_gr object}
# Select a plot region
plot_region <- "chr1:1156819-1178981"

# Extract cSEAdb information
search_db_table <- search_db(plot_region,
                             query_type="se_region",se_db=cSEAdb)

# User can also add other H3K27Ac bigwig files beyond the NCI-60 cell panel
usr_bigwig <- system.file("extdata","example_signal.bw",package="cSEAdb")

# Create bigwig GRange object for downstream plotting
bw_gr <- create_bw_gr(plot_region,c(bw_path,usr_bigwig),se_spec_table=search_db_table)
```

Next, users can filter or specify which cell lines to be plotted. For demo purpose, we only look into cell lines that the queried SEs show specificity (*specific*). We also manually add other cell lines (A549, MCF7) and a customized sample. 

```{r create tracks}

# add hg38 genomes for gene tracks (optional)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

# Create plot object for gviz package with gene tracks
plot_object <- create_gviz_tracks(bw_gr_list=bw_gr$bw_gr,
                               cell="specific,A549,MCF7,example_signal.bw",
                               se_spec_table=search_db_table,
                               plot_region=plot_region,
                               txdb=TxDb.Hsapiens.UCSC.hg38.knownGene)

# Names of plot_object elements
names(plot_object)
```

The data object includes the following 7 types of information to be plotted.
Users can also customize which tracks to be plotted using Gviz package.

  1. htrack: main bigwig tracks (must include).
  1. atrack_ce: CE regions.
  1. atrack_se: SE regions.
  1. atrack_query: query regions.
  1. gtrack: genomic tracks with gene information
  1. plot_region: regions to be plotted.
  1. axistrack: genome axis track indicating locations.
  

## Plot all tracks

Example to visualize with all types of information in the data object.

```{r track1, fig.height=20, fig.width=15}
# with the plotting object, users can customize which track to plot using gviz package
plotTracks(c(plot_object$htrack,
             plot_object$atrack_ce,
             plot_object$atrack_se,
             plot_object$atrack_query,
             plot_object$axistrack,
             plot_object$gtrack),
           size=10,
           cex.title = 0.8)

```

## Without annotation tracks

Example	to visualize with the selected types of information in the data object.

```{r track2, fig.height=20, fig.width=15}
# with the plotting object, users can customize which track to plot using gviz package
plotTracks(c(plot_object$htrack,
             plot_object$atrack_ce,
             plot_object$atrack_se),
           size=10,
           cex.title = 1)

```

# Generate chromosome maps

In this section, we demo how to generate chromosome maps of specific SEs/CEs using data from A549 cell line.

```{r chromosome maps}
# load tools
library(chromoMap)

# read cell spec chromosome file and chromosome size file
cell_spec <- readRDS(system.file("data","cell_spec_for_chr_plot.rds",package="cSEAdb"))
chrom_size <- read.table(system.file("extdata","hg38_chrom_loc_w_centromere.txt",package="cSEAdb"),
                         sep="\t",header=F)
head(chrom_size)

# Extract specificity information with A549
cell_spec_example <- cell_spec$A549

# Generate chromosome map with chromoMap package
chr_map <- chromoMap(list(chrom_size),list(cell_spec_example),
                          n_win.factor = 2,
                          # chr_width = 20,
                          chr_length = 6,
                          chr_color = "grey",
                          data_based_color_map = T,
                          data_type = "categorical",
                          legend = T,
                          lg_x = 200,
                          lg_y = 500,
                          title="A549",
                          text_font_size = c(16),
                          data_colors = list(c("firebrick3","lightskyblue","gold3")))

chr_map
```
