---
title: "_cSEAdb_ User's Guide"
author: |
  | Xiang Liu <xiang.liu@moffitt.org>
  | Mingxiang Teng <mingxiang.teng@moffitt.org>
  | Department of Biostatistics and Bioinformatics
  | Moffit Cancer Center, Tampa, FL, USA
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
  pdf_document:
    toc: yes
graphics: yes
vignette: |
  %\VignetteIndexEntry{cSEAdb user's guide} %\VignettePackage{cSEAdb} %\VignetteEncoding{UTF-8} %\VignetteEngine{knitr::rmarkdown}
---

```{r, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  error = FALSE,
  warning = FALSE, 
  message = FALSE,
  cache = FALSE,
  comment = "#>",
  dev="png"
)
```

# Introduction
Super enhancers (SE) have been found with cancer-specific gene regulation in human cancers. Recent studies reported the importance of understanding the cooperation and function of their internal components, i.e., the constituent enhancers (CE). However, there are no pan-cancer studies to identify cancer-specific SE signatures at the constituent level. Here, we report fingerprint SE signatures for 28 cancer types in the NCI-60 cell panel by implementing a mixture model to automatically discriminate active CEs from inactive CEs. This R package cSEAdb was built to facilitate the query, exploration, and visualization of the fingerprint SEs in NCI-60 cell lines.

# Install package

Install _cSEAdb_ package via _devtools_.
```{r install}
library(devtools)
devtools::install_github("https://github.com/tenglab/cSEAdb.git")
```

# Getting Started

Load the package in R

```{r library}
library(cSEAdb)
```

# Read cSEAdb RDS object

```{r readRDS}
cSEAdb <- readRDS(system.file("data","cSEAdb.rds",package="cSEAdb"))
names(cSEAdb)
```

# Tables of cSEAdb object

cSEAdb includes the following files:

  1. se_specificity: main fingerprint file including SE and CE cell and cancer specificity. Meaning of each columns are as follow:
    + se_name: SE names (chr_start_end).
    + number_spec_ce: number of specific CEs each SE have.
    + number_spec_object: number of cell (cancer) which each CE are specific.
    + spec_ce: CEs which have cell or cancer specificity.
    + spec_object: cell line for cell-specific; cancer name for cancer-specific.
    + specificity: cell(cancer) specificity of each CE; active, inactive or non-specific.
    + object_type: cell-specific or cancer-specific.
  1. se_bed: Super-enhancer BED file, a merged SE BED file from 60 cell lines.
  1. ce_bed: a merged constituent enhancer (CE) BED file from 60 cell lines.
  1. cell_cancer_meta: metadata of cell line and cancers of 60 cell lines.
  1. ce_signal: normalized coverage signal of all CEs.
  1. ce_mixture_model_cell: mixture group code of all CEs (low signal group: 0; high signal group: 1).
  1. gene_promotor: gene promotor regions based on hg38 used for gene query.
  

# Query with _cSEAdb_

In this section, we extract specificity information of SE and CE with search_db function using different query types.

### *Query _cSEAdb_ with SE regions*

```{r Query SE regions}
# regions
hits_1 <- search_db(c("chr1_100584859_100673000","chr1_112388326_112392055"),
                     query_type="se_region",se_db=cSEAdb)
head(hits_1)

# bed file
bed_file <- system.file("extdata","se_query_example.bed",package="cSEAdb")
hits_2 <- search_db(bed_file,query_type="se_bed",se_db=cSEAdb)

head(hits_2)
```

### *Query _cSEAdb_ with gene names*

```{r Query with gene names}
hits_gene <- search_db(query=c("MYC","FOXA2"),
                        query_type="gene",se_db=cSEAdb)
head(hits_gene)
```

### *Query _cSEAdb_ with cell lines (only for the 60 cell line in this study)*

```{r Query with cell lines}
# cell line names used in this study
cell_name <- unique(cSEAdb$cell_cancer_meta$Cell_line)
head(cell_name)

# search cSEAdb with cell name
hits_cell <- search_db(cell_name[c(1,2)],
                       query_type="cell",se_db=cSEAdb)
head(hits_cell)
```

### *Query _cSEAdb_ with cancers (only for the 28 cancer types in this study)*

```{r Query with cancer names}
# cancer names used in this study
cancer_name <- unique(cSEAdb$cell_cancer_meta$Cell_line)
head(cancer_name)

# search cSEAdb with cancer names
hits_cancer <- search_db(query=cancer_name[c(1,2)],
                       query_type="cancer",cSEAdb)
head(hits_cancer)
```

# Generate signal track plot

This section demostrates how to generate track plots with a genomic regions.

```{r download bigwig files of 60 cell lines}
# Download the normalized bigwig file of 60 cell lines
# Download link are located in data/bw_all_plinks.txt file
bw_links <- system.file("extdata","bw_all_plinks.txt",package="cSEAdb")
bw_download_link <- read.table(bw_links,sep="\t",header=F)
head(bw_download_link)

# Create a bigwig list with the path of bigwig files
# Here is a example of path list of downloaded bigwig files
bw_dir <- system.file("extdata","example_bw",package="cSEAdb")
bw_path  <- list.files(bw_dir,full.names = T)
head(bw_path)
```

Once we have the bigwig path, we can create a Grange object with region like to plot.

```{r create bw_gr object}
# Define a plot region
plot_region <- "chr1:1156819-1178981"

# Extract cSEAdb information
search_db_table <- search_db(plot_region,
                             query_type="se_region",se_db=cSEAdb)

# You can add your own bigwig files
usr_bigwig <- system.file("extdata","example_signal.bw",package="cSEAdb")

# Create bigwig gr object with function create_bw_gr
bw_gr <- create_bw_gr(plot_region,c(bw_path,usr_bigwig),se_spec_table=search_db_table)
```

After generate the bigwig Granges, we can create gnomic tracks. We can chose any cell lines provided in this database. Here we only generate the cells which have specificity in this region. We also add some extra cells (A549, MCF7) and user's own samples. 

```{r create tracks}

# add hg38 genomes for gene tracks (optional)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

# Create plot object for gviz package with gene tracks
plot_object <- create_gviz_tracks(bw_gr_list=bw_gr$bw_gr,
                               cell="specific,A549,MCF7,example_signal.bw",
                               se_spec_table=search_db_table,
                               plot_region=plot_region,
                               txdb=TxDb.Hsapiens.UCSC.hg38.knownGene)

# Names of plot_object elements
names(plot_object)
```

The plot object includes the following files:

  1. htrack: main bigwig tracks (must include).
  1. atrack_ce: CE regions.
  1. atrack_se: SE regions.
  1. atrack_query: query regions.
  1. gtrack: genomic track with gene information
  1. plot_region: regions to plot.
  1. axistrack: genome axis track indicate locations.
  
Then users can customize which track to plot using Gviz package.

## Plot all tracks

```{r track1}
# with the plot object, users can customize which track to plot using gviz package
plotTracks(c(plot_object$htrack,
             plot_object$atrack_ce,
             plot_object$atrack_se,
             plot_object$atrack_query,
             plot_object$axistrack,
             plot_object$gtrack),
           cex.title = 0.4)

```

## Without annotation tracks

```{r track2}
# with the plot object, users can customize which track to plot using gviz package
plotTracks(c(plot_object$htrack,
             plot_object$atrack_ce,
             plot_object$atrack_se),
           cex.title = 0.4)

```

# Generate chromosome maps

This section demonstrates how to generate chromosome maps with the information extracted from _cSEAdb_. _cSEAdb_ includes cell and cancer RDS object for chromosome plots. Here we use A549 as an example.

```{r chromosome maps}
# load tools
library(chromoMap)

# read cell spec chromosome file and chromosome size file
cell_spec <- readRDS(system.file("data","cell_spec_for_chr_plot.rds",package="cSEAdb"))
chrom_size <- read.table(system.file("extdata","hg38_chrom_loc_w_centromere.txt",package="cSEAdb"),
                         sep="\t",header=F)
head(chrom_size)

# Extract specificity information with A549
cell_spec_example <- cell_spec$A549

# Generate chromosome map with chromoMap package
chr_map <- chromoMap(list(chrom_size),list(cell_spec_example),
                          n_win.factor = 2,
                          # chr_width = 20,
                          chr_length = 6,
                          chr_color = "grey",
                          data_based_color_map = T,
                          data_type = "categorical",
                          legend = T,
                          lg_x = 200,
                          lg_y = 500,
                          title="A549",
                          text_font_size = c(16),
                          data_colors = list(c("firebrick3","lightskyblue","gold3")))

chr_map
```
